### PR TITLE
Use the ``tomli`` package instead of ``toml`` to parse ``.toml``

### LINKED ISSUE #5885: Replace `toml` with `tomli`
PEP 680 was accepted recently. Starting with 3.11, toml parsing will be part of the stdlib. To prepare for that, we should switch from the current dependency `toml`, to `tomli>2.0` which is the basis for the stdlib implementation.

A good example of how this can be implemented can be found here: https://github.com/python/mypy/pull/12305

**Some useful resources**
https://docs.python.org/3.11/library/tomllib.html#module-tomllib
https://peps.python.org/pep-0680/
https://github.com/hukkin/tomli

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Closes #5885.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -51,7 +51,7 @@ install_requires =
     astroid>=2.9.2,<=2.10.0-dev0
     isort>=4.2.5,<6
     mccabe>=0.6,<0.7
-    toml>=0.9.2
+    tomli>=1.1.0;python_version"3.11"
```
Comment (DanielNoord) [inline setup.cfg:54]:
```suggestion
    tomli>=1.1.0;python_version<"3.11"
```

File: tests/config/functional/toml/issue_3181/toml_decode_error.1.out [hunk 2]
```diff
@@ -1,2 +1,2 @@
 ************* Module {abspath}
-{relpath}:1:0: F0011: error while parsing the configuration: Found invalid character in key name: '*'. Try quoting the key name. (line 3 column 2 char 48) (config-parse-error)
+{relpath}:1:0: F0011: error while parsing the configuration: Invalid statement (at line 3, column 1) (config-parse-error)
```
Comment (Pierre-Sassoulas) [inline tests/config/functional/toml/issue_3181/toml_decode_error.1.out:2]: Well, it's a little sad to see the default builtin lib have worst error messages by far than the old one.

**Review by Pierre-Sassoulas — APPROVED** [review]
Great, thank you ! I think we could add ``3.11`` to github action before merging this ?

**Comment by DanielNoord** [comment]
> Great, thank you ! I think we could add `3.11` to github action before merging this ?

I believe there were issues with turning that on previously. I'd save that for another time.

File: pylint/config/find_default_config_files.py [hunk 3]
```diff
@@ -3,17 +3,20 @@
...
-from toml import TomlDecodeError
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib
 
 
 def _toml_has_config(path):
-    with open(path, encoding="utf-8") as toml_handle:
+    with open(path, mode="r+b") as toml_handle:
```
Comment (cdce8p) [inline pylint/config/find_default_config_files.py:16]:
```suggestion
    with open(path, mode="rb") as toml_handle:
```

File: pylint/config/option_manager_mixin.py [hunk 4]
```diff
@@ -300,8 +303,8 @@ def _parse_toml(
         self, config_file: Union[Path, str], parser: configparser.ConfigParser
     ) -> None:
         """Parse and handle errors of a toml configuration file."""
-        with open(config_file, encoding="utf-8") as fp:
-            content = toml.load(fp)
+        with open(config_file, mode="r+b") as fp:
```
Comment (cdce8p) [inline pylint/config/option_manager_mixin.py:306]:
```suggestion
        with open(config_file, mode="rb") as fp:
```

**Review by cdce8p — APPROVED** [review]
Looks good. Thanks @DanielNoord 🐬

**Pull request merged**